core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1125,6 +1142,78 @@ private[spark] class TaskSetManager(
   def executorAdded(): Unit = {
     recomputeLocality()
   }
+
+  /**
+   * A class for checking inefficient tasks to be speculated, the inefficient tasks come from
+   * the tasks which may be speculated by the previous strategy.
+   */
+  private class InefficientTask {
+    private var taskData: Map[Long, TaskData] = null
+    private var successTaskProgress = 0.0
+    private val checkInefficientTask = speculationTaskMinDuration > 0
+
+    if (checkInefficientTask) {
+      val appStatusStore = sched.sc.statusTracker.getAppStatusStore
+      if (appStatusStore != null) {
+        successTaskProgress =
+          computeSuccessTaskProgress(taskSet.stageId, taskSet.stageAttemptId, appStatusStore)
+        val stageData = appStatusStore.stageAttempt(taskSet.stageId, taskSet.stageAttemptId, true)
+        if (stageData != null) {
+          taskData = stageData._1.tasks.orNull
+        }
+      }
+    }
+
+    private def computeSuccessTaskProgress(stageId: Int, stageAttemptId: Int,
+      appStatusStore: AppStatusStore): Double = {
+      var sumInputRecords, sumShuffleReadRecords, sumExecutorRunTime = 0.0
+      appStatusStore.taskList(stageId, stageAttemptId, Int.MaxValue).filter {
+        _.status == "SUCCESS"
+      }.map(_.taskMetrics).filter(_.isDefined).map(_.get).foreach { task =>
+        if (task.inputMetrics != null) {
+          sumInputRecords += task.inputMetrics.recordsRead
+        }
```

Comment:
how about recordsWritten? Should that also be considered wrt progress same wrt shuffleRecordsWritten?

Reply:
Even cache can also take time when written to disk, does that need to be taken into consideration? Similarly GC time, shuffle read blocked time etc. could also impact task progress